[draft] add tw asana

maxenglander opened the pull request:
# Description

New, rudimentary sync service for TW<->Asana.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

With manual tests so far. Will add automated tests, but would like some initial high-level feedback.

## Setup

1. Create an Asana account
2. Generate an Asana Personal Access Token
3. Store Personal Access Token in password store
4. Find Asana workspace GID with `tw_asana_sync --list-asana-workspaces`

## Tests

Do these tests in-order.

- [x] Create task on Asana in workspace <workspace-gid>, sync to TW
   1. Create a task on Asana
   2. Run `tw_asana_sync --asana-workspace <workspace-gid> -t test-asana`
   3. Verify task is created in Asana with name of the Asana task set to the description of the TW task
   4. Verify and take note of a new combination is that save, it should look something like `tw_asana_sync --combination <workspace-gid>__None__test-asana`
- [x] Modify task on Asana sync to TW
   1. Modify description of last task, changing the name and due date
   2. Run the newly created combination `tw_asana_sync --combination <workspace-gid>__None__test-asana`
   3. Verify task is modified in TW, with updated `description` and a `due` fields
- [x] After no further changes, sync detects no changes
   1. Run `tw_asana_sync --asana-workspace <workspace-gid> -t test-asana`
   2. Verify no changes are detected or applied to either Asana or TW
- [x] Modify task in TW sync to Asana
   1. In TW, Change `description` and remove `due` fields of last task 
   3. Run `tw_asana_sync --asana-workspace <workspace-gid> -t test-asana`
   4. Verify task is modified in Asana, with updated name removed due date
- [x] After no further changes, sync detects no changes
   1. Run `tw_asana_sync --asana-workspace <workspace-gid> -t test-asana`
   2. Verify no changes are detected or applied to either Asana or TW
- [x] Delete task in Asana, verify deleted in TW
   1. Delete the last task in Asana
   2. Run `tw_asana_sync --asana-workspace <workspace-gid> -t test-asana`
   3. Verify the task is marked as deleted (but not purged) in TW
- [x] Create untagged task in TW, is not synced to Asana
   1. Create a task in TW without the `test-asana` tag
   2.  Run `tw_asana_sync --asana-workspace <workspace-gid> -t test-asana`
   3. Verify task is not created in Asana
- [x] Add tag and due date to TW task, sync to Asana
   1. Add `test-asana` tag and a due date to last task
   2.  Run `tw_asana_sync --asana-workspace <workspace-gid> -t test-asana`
   3. Verify task is created in Asana with a due date
- [x] Remove TW due date, sync to Asana
   1. Remove due date from Asana task
   2.  Run `tw_asana_sync --asana-workspace <workspace-gid> -t test-asana`
   3. Verify due date is removed from Asana task
- [x] After no further changes, sync detects no changes
   1. Run `tw_asana_sync --asana-workspace <workspace-gid> -t test-asana`
   2. Verify no changes are detected or applied to either Asana or TW
- [x] Complete TW task, sync to Asana
   1. Mark last task as completed in TW
   3.  Run `tw_asana_sync --asana-workspace <workspace-gid> -t test-asana`
   5. Verify task is marked completed in Asana
- [x] Un-complete task in Asana, sync to TW
   1. Mark last task as incomplete in Asana
   2.  Run `tw_asana_sync --asana-workspace <workspace-gid> -t test-asana`
   3. Verify task is marked incomlete in TW
- [x] Delete task in TW, sync to Asana
   1.  Delete task in TW
   2.  Run `tw_asana_sync --asana-workspace <workspace-gid> -t test-asana`
   3. Verify task is deleted (non-permanently) in Asana
    

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been merged and published in downstream modules
       ^ None were needed

bergercookie commented:
Great stuff @maxenglander, thanks for the contribution!

let me go through the changes in the next few days and I'll give you more concrete feedback then :).

All in all, it looks great though!

bergercookie commented on readme-asana.md:
```diff
@@ -0,0 +1,67 @@
+Current limitations:
+
+- Does not sync Asana tags, project names, subtasks, projects, likes, etc.
+- Does not sync Taskwarrior tags, projects, annotations, etc.
```
this is probably outdated? Syncing taskwarrior tags works fine for me

maxenglander replied:
Oh really? That's cool!

maxenglander replied:
I tried this and it seemed to me like tags were not synced. I added tags to an Asana task, did an import, and they didn't show up on the TW side.

bergercookie replied:
Sorry, probably I didn't express this right. I meant that you can use `-t <some-tw-tag` to only sync tasks of a certain TW tag with Asana.

```
Usage: tw_asana_sync [OPTIONS]

Options:
  --token, --token-pass-path TEXT
                                  Path in the UNIX password manager to fetch
  -w, --asana-workspace-gid TEXT  Asana workspace GID used to filter tasks
  -W, --asana-workspace-name TEXT
                                  Asana workspace name used to filter tasks
  --list-asana-workspaces         List the available Asana workspaces
  -t, --taskwarrior-tags TEXT     Taskwarrior tags to sync <----
```

Isn't this supported ?

bergercookie commented on readme-asana.md:
````diff
@@ -0,0 +1,67 @@
+To synchronize Asana tasks within a workspace:
+
+```sh
+tw_asana_sync --tags asana --asana-workspace 123456789012345
````
nitpick: Maybe we should allow the user to sync a workspace by specifying either its name or GID?

maxenglander replied:
We can do that. Unfortunately there's nothing prevent the user from having duplicate workspace names, but that's probably not a common situation. We can default to mapping name to GID, and if we see that there's ambiguity, exit and ask user to select GID?

bergercookie replied:
>  We can default to mapping name to GID, and if we see that there's ambiguity, exit and ask user to select GID?

Sounds like a good alternative.. but again, not too fussed about it if it's too much work

maxenglander replied:
Made changes in latest commit so that either (but not both) `--asana-task-gid` or `--asana-task-name` are accepted. If `--asana-task-name` is ambiguous, a `RuntimeError` is raised.

bergercookie commented on taskwarrior_syncall/cli.py:
```diff
@@ -9,6 +9,46 @@
+        "--token-pass-path",
+        "token_pass_path",
+        help="Path in the UNIX password manager to fetch",
+        default="app.asana.com/0/my-apps/taskwarrior_syncall",
```
Can we remove this default option? Each one ofthe users will have their own password layout so mandating one here wouldn't really work. Also feel free to remove the default from the notion flag [here](https://github.com/bergercookie/taskwarrior-syncall/pull/52/files#diff-eb15d8a04ad5649425284945bae54ca17bb2984ed69827769aed17b454dbf8b7R131)

maxenglander replied:
Will do

maxenglander replied:
Done in latest commit

bergercookie commented on taskwarrior_syncall/scripts/tw_asana_sync.py:
```diff
@@ -0,0 +1,226 @@
+    # and get all the tasks in each workspace. For now,
+    # require user to specify the workspace to pull tasks
+    # from.
+    assert asana_workspace_gid or asana_task_gid
```
Instead of asserting, would it be better to check and raise a `RuntimeError` with a descriptive error message?

maxenglander replied:
Yes :) Will do

maxenglander replied:
Done in latest commit

bergercookie commented on taskwarrior_syncall/tw_asana_utils.py:
```diff
@@ -0,0 +1,160 @@
+    """List the Asana workspaces known to the provided client."""
+    items = {}
+
+    workspaces = client.workspaces.find_all()
```
See my comment in the PR main page but iterating this seems to make an API call per item which doesnt' really scale (will take a long while to complete if the user has 100s of items in Asana.

* Is there a client method to fetch these in bulk - maybe setting the `page_size` to a much higher value might help? https://github.com/Asana/python-asana/#options=

maxenglander replied:
Thanks for calling this out.

> but iterating this seems to make an API call per item

This is actually intentional.

https://github.com/bergercookie/taskwarrior-syncall/pull/52/files#diff-e32b17c14d925106b95e4bc6e38235a0fcee593ce06f14948d89eaf08e7e3177R35

The Asana API gives us high-level task info with `.find_all()`, but in order to get details like `due_at` we need to make follow-up API call for each task.

I will increase `page_size` because that's a good idea, but I'm not sure how much faster it will make things.

Some thoughts on other things we can do to improve....

We can make use of API filters `--completed-since` and `--modified-since`. For example:

 1. The first time we sync a combination, we can use `--completed-since=now`, which will only return incomplete tasks. 
     * Maybe the user wants to pull down completed tasks too, in which case we can give users a flag to opt-in to that
 2. Subsequent syncs can use the `--modified-since=<last-sync-time>` API
 3. Every sync will trigger saving `<last-sync-time>` to disk somewhere.

The one major issue with this idea is that the current framework will delete items if they aren't returned by `.get_items()`. Probably a simple way to deal with this would be to add a `--ignore-asana-deletes` option to the framework not to process deletes from the Asana side. This isn't ideal, but users could mitigate this by doing a periodic full sync.

bergercookie replied:
> We can make use of API filters --completed-since and --modified-since.

Yeah, agree, both to this and then to defaulting to the modified time 👍🏾 

> Maybe the user wants to pull down completed tasks too, in which case we can give users a flag to opt-in to that

Agree, let's do this!

>  The one major issue with this idea is that the current framework will delete items if they aren't returned by .get_items(). Probably a simple way to deal with this would be to add a --ignore-asana-delete

Sounds good..
Indeed, currently `item_synchronizer` it will flag it as deleted if it's not returned by the `get_all_items`, which isn't always what one may want. This seems like a functionality that other sync scripts may want, so when adding the flag for `--ignore-asana-deletes`, I'd create a more generic function that accepts the name of the side that it wants to create the flag for, similar to, `opt_combination(name_A: str, name_B: str)` but with a single parameter instead of two.

maxenglander replied:
I haven't reworked how tasks are fetched, and if it's alright with you will left that out of this PR, but I did raise the page size from 50 to 100 (the max size allowed by the API).

bergercookie commented:
Hi @maxenglander , do you have some time to wrap this PR?

Let me know when you think it's in a ready state and I can merge. I can deal with the page_size and making fetching the tasks faster at a later PR, but for now I'd like to get it merged since it's a very cool feature to add :)

maxenglander commented:
Hey @bergercookie thanks for pinging. I think I can get this wrapped up in the next week! Thanks for being flexible on what needs to be addressed in this PR. Will ping you when it's ready for another look.

maxenglander commented:
Hey @bergercookie I addressed some of your PR comments. I didn't rework how tasks are fetched for now, but addressed your other comments. I want to do another round of local tests, and then I think this will be good to go.

maxenglander commented:
OK @bergercookie think it's ready for another look :)

bergercookie commented:
Hi @maxenglander , great stuff! I'll have a thorough look at it by the weekend :)

bergercookie commented:
It works great!

I'm happy to merge this once the CI is green. Seems that it's currently failing. Let me know if you need any help on this.

https://github.com/bergercookie/taskwarrior-syncall/runs/8291921377?check_suite_focus=true

Also, some non-functionality related items to wrap this up. Hopefully this shouldn't take long:

* Could you add a line in the following file to generate shell completion for the `tw_asana_sync` script? https://github.com/maxenglander/taskwarrior-syncall/blob/add-tw-asana/scripts/generate-shell-completions.sh#L57
* Could you perhaps include a screenshot example of creating the asana API token in the README -  just to make it a bit clearer? Not important. We can skip this if you think it's not needed.
* Could you rebase on top of latest master squash and cleanup the git hisotory?

bergercookie commented:
Failing just the coverage check which is a bit flaky anyway. Merging this 🎉

bergercookie commented:
@maxenglander thanks for your work!
I took the liberty to fix the failing test and update the readme (see the last commit) since I had some free time today!  I'll deploy it to pypi ASAP too.

Thanks again! Great stuff! 🎉🎉🎉

maxenglander commented:
@bergercookie thanks for getting this across the finish line! Appreciate the support throughout and am happy to see it merged :)

bergercookie commented:
For future reference, Asana should be available in versions >= 1.3.0 in PyPI
